File: 이상욱/BOJ/step/BOJ9012.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class BOJ9012 {
	public static void main(String[] args) throws NumberFormatException, IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());

		for (int i = 0; i < n; i++) {
			String vps = br.readLine();
			String answer = check(vps);
			System.out.println(answer);
		}
	}

	private static String check(String vps) {

		Stack<Character> stk = new Stack<>();

		for (int i = 0; i < vps.length(); i++) {

			if (stk.isEmpty()) {
				stk.add(vps.charAt(i));
			} else {

				char c = stk.peek();

				if (c == '(' && vps.charAt(i) == ')') {
					stk.pop();

				} else {
					stk.add(vps.charAt(i));
				}

			}

		}

		if (stk.isEmpty()) {
			return "YES";
		}

		else {
			return "NO";
		}

	}
}

File: 이상욱/BOJ/step/BOJ9012.kt
package level17

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main(){

    val br = BufferedReader(InputStreamReader(System.`in`))

    val n = br.readLine().toInt()

    repeat(n){

        val vps = br.readLine()
        val answer = check(vps)
        println(answer)

    }

}

fun check(vps : String) : String{

    val stk = Stack<Char>()

    for(i in vps.indices){

        if(stk.isEmpty()){
            stk.add(vps[i])
        }else{
            val c = stk.peek()

            if(c == '(' && vps[i] == ')'){
                stk.pop()
            }else{
                stk.add(vps[i])
            }
        }

    }

    return if(stk.isEmpty()){
        "YES"
    }else{
        "NO"
    }

}